Improve numerical stability of GroupNorm

xiaomengy opened the pull request:
Summary: Improve numerical stability of GroupNorm

Test Plan: buck test mode/dev-nosan //caffe2/test:nn -- "GroupNorm"

Differential Revision: D27414438

There are two main changes in this PR:
1. Use Welford algorithm to compute main and variance.
2. Use cascade summation when computing sum over input for both mean and variance.

Some benchmark results.

input_size = [256, 1024, 28, 28], group = 8

for single thread CPU run average time change is 510.91ms -> 602.70ms.
for CUDA run average time change is 12.20ms -> 12.47ms.

ngimel commented on aten/src/ATen/native/cpu/moments_utils.h:
```diff
@@ -0,0 +1,129 @@
+
+  constexpr int64_t kVecSize = Vec::size();
+  constexpr int64_t kChunkSize = 32;
+  constexpr int64_t kMaxDepth = 64;
```
max depth of 64 seems like an overkill (given that it is then used as the size for the m0_stk, m1_stk, m2_stk variables)? In the reduction PR #39516 there were just 4 levels and that seemed enough for pretty large tensors.

xiaomengy replied:
Personally I think it will be more reliable to keep a fixed chunk size, especially for the case in var because there is square op for the input. I changed the chunk size to 128 which is the same as numpy below. And also make the kMaxDepth as a parameter for the template function. Then if the size is small, the SmallVector will be smaller to improve the register efficiency.
https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/core/src/umath/loops_utils.h.src#L60

peterbell10 replied:
I chose a fixed number of levels because I expect the compiler to do a lot better at allocating registers for a fixed number of accumulators. I don't think a dynamically-sized array of vector is even possible to keep in registers.

For four levels with a chunk size of 16, and a vector size of 8 (for float), you can sum up to 2^19 or .5 million elements. With 5 levels and a chunk size of 16, you can sum 2^63 elements i.e. more floats than you can possibly address with 64-bits.

With 64 levels and a chunk size of 16, you could count every atom in the universe.

ngimel commented on aten/src/ATen/native/cpu/SumKernel.cpp:
```diff
@@ -1,11 +1,12 @@
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/cpu/Reduce.h>
+#include <ATen/native/cpu/utils.h>
+#include <c10/util/llvmMathExtras.h>
```
nit: looks like llvmMathExtras are not needed here, only by cpu/utils.h?

xiaomengy replied:
Done.

ngimel commented on aten/src/ATen/native/cuda/block_reduce.cuh:
```diff
@@ -34,6 +36,55 @@ __inline__ __device__ T BlockReduceSum(T val, T* shared) {
 }
 
+template <typename T>
+__inline__ __device__ thrust::tuple<int64_t, T, T> WarpReduceMoments(
```
is there any reason you are not using WelfordOps.combine from SharedReduceOps.h here?

xiaomengy replied:
Changed to WelfordOps

ngimel commented on torch/testing/_internal/common_nn.py:
```diff
@@ -1674,6 +1674,19 @@ def fractional_max_pool3d_test(test_case):
+        desc='2d_no_affine_large_feature_forward',
+        check_forward_only=True,
+        reference_fn=lambda x, *_: torch.nn.functional.layer_norm(
+            x.view(4, 3, -1), (4 * 28 * 28,), eps=1e-3).view(4, 12, 28, 28)
```
is this size enough to observe better accuracy (i.e. will it fail without this PR?). Also, IIRC you never landed LayerNorm better accuracy PRs, so using LayerNorm as reference could be inaccurate?

xiaomengy replied:
The original purpose of this test is to make sure we cover the test cases larger than the chunksize. It is a little hard to have a reliable test case for the numerical issue. If the input size is large, the test will be pretty slow. I removed the layer_norm and used torch.mean and torch.std for reference_fn.

ngimel reviewed (commented):
This looks good!
Can you please post benchmarks, both for cuda and cpu?
Also, consider implementing multi-stage cascade summation, like in original `sum` PR or #55841 to limit number of levels (64 is a bit too much), but it's not a blocker, PR can go in as is.

peterbell10 commented on aten/src/ATen/native/cpu/moments_utils.h:
```diff
@@ -0,0 +1,146 @@
+      m2_vec += delta_vec * (x_vec - m1_vec);
+    }
+    AddMomentsVec(m0, m1_vec, m2_vec, m0_stk[0], m1_stk[0], m2_stk[0]);
+    int64_t mask = i + 1;
+    for (int64_t j = 1; j < depth && (mask & 1) == 0; ++j) {
```
This isn't right. Each level of the accumulator should be the sum of `kChunkSize` elements on the level below.

xiaomengy replied:
This is the correct formula. Here I only do the chunk_size reduction on the leaf node. The reduction tree is a binary tree. The pseudo code is the same as numpy's implementation below. Here I think it will be more reliable to do the exact pairwise summation for var since it is actually the 2nd order moment, which will enlarge the numeric issue for summation.
https://github.com/numpy/numpy/blob/623bc1fae1d47df24e7f1e29321d0c0ba2771ce0/numpy/core/src/umath/loops_utils.h.src#L135

xiaomengy commented on aten/src/ATen/native/SharedReduceOps.h:
```diff
@@ -80,8 +80,20 @@ struct WelfordData {
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  // Shared memory requires no initialization.
+  C10_HOST_DEVICE WelfordData() = default;
```
Yes, but if we add dynamic initialization in ctor here, there will be a warning "dynamic initialization is not supported for a function-scope static __shared__ variable within a __device__/__global__ function".

ngimel replied:
Sorry for delay, I'm still uneasy about this initialization, otherwise PR looks good.

xiaomengy replied:
Actually there are some discussions for this before. Not sure if there is any better way to handle this.
https://github.com/pytorch/pytorch/pull/13967

zasdfgbnm replied:
Quoting from our compiler team for a similar issue:

```
__shared__ variables cannot be initialized.
```
...
and for the case of dynamic initialization:
```
the compiler reports a warning (rather than an error, for legacy reasons). The program will build but will not execute correctly. The warning reported is:

j5.cu(6): warning: dynamic initialization is not supported for a function-scope static __shared__ variable within a __device__/__global__ function
```

Can we keep the default ctor unchanged as
```
C10_HOST_DEVICE WelfordData() : mean(0), m2(0), n(0), nf(0)  {}
```

and inside the kernel, do
```CUDA
__shared__ char val_shared_[C10_WARP_SIZE * sizeof(WelfordType)];
WelfordType *val_shared = reinterpret_cast<WelfordType *>(val_shared_);
if (threadIdx.x < 32) {
  val_shared[threadIdx.x] = WelfordType();
}
__syncthreads();
```

xiaomengy replied:
Thanks for the advice.

xiaomengy reviewed (commented):
Here I think it will be more reliable to do the exact pairwise summation for var since it is actually the 2nd order moment, which will enlarge the numeric issue for summation. So I didn't keep a fixed size of depth. Since there is only the chunk-size reduction on the leaf node, so the actual max depth should be 63 - (4 + 3) = 56. So I think the maximum of 64 should be reasonable. Though in practice we will never use that size.

ngimel commented on aten/src/ATen/native/cuda/group_norm_kernel.cu:
```diff
@@ -32,29 +32,36 @@ __global__ void RowwiseMomentsCUDAKernel(
-    sum2 = cuda_utils::BlockReduceSum<T_ACC>(sum2, v_shared);
+    // There will be a warning if we declare a __shared__ WelfordType array.
+    // https://github.com/pytorch/pytorch/pull/13967
+    __shared__ char val_shared[C10_WARP_SIZE * sizeof(WelfordType)];
```
this can potentially result in misaligned access, so please use 
```
__shared__ std::aligned_storage<sizeof(WelfordType), alignof(WelfordType)>::type val_shared
```

xiaomengy replied:
Thanks for the suggestions!

ngimel commented on torch/testing/_internal/common_nn.py:
```diff
@@ -1647,6 +1647,20 @@ def fractional_max_pool3d_test(test_case):
+        module_name='GroupNorm',
+        constructor_args=(3, 48, 0.0),
+        cpp_constructor_args='torch::nn::GroupNormOptions(3, 48).eps(0.0)',
+        input_fn=lambda: torch.rand(2, 48, 15, 15),
```
Do you think this size is enough to check accuracy?

xiaomengy replied:
A smoke test case for large input added.

ngimel reviewed (commented):
This looks good, I have a minor comment about alignment.

ngimel commented on test/test_nn.py:
```diff
@@ -5760,6 +5760,38 @@ def test_RNN_cell_forward_hidden_size(self):
                      nn.GRUCell(*cell_shared_param)):
             self.assertRaises(Exception, lambda: cell(input, hx))
 
+    def test_GroupNorm_numeric(self):
```
can you move this test into device-generic part of the test suite (TestNNDeviceType at line ~11568) so that you'd have device argument to handle cuda testing?
How long does the cpu test take? Should it be decorated with @slowTest?

xiaomengy replied:
Done.

xiaomengy committed:
Improve numerical stability of GroupNorm (#54921)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/54921

Improve numerical stability of GroupNorm

Test Plan: buck test mode/dev-nosan //caffe2/test:nn -- "GroupNorm"

Reviewed By: ngimel

Differential Revision: D27414438

fbshipit-source-id: 99180a4f8d571c78bea1c4b58cd4b7811bf60ae5
